mono/metadata/native-library.c

```diff
@@ -947,3 +1370,145 @@ pinvoke_probe_for_symbol (MonoDl *module, MonoMethodPInvoke *piinfo, const char
 	*error_msg_out = error_msg;
 	return addr;
 }
+
+#ifdef ENABLE_NETCORE
+void
+ves_icall_System_Runtime_InteropServices_NativeLibrary_FreeLib (gpointer lib, MonoError *error)
+{
+	MonoDl *module;
+	guint32 ref_count;
+
+	// Don't free __Internal
+	if (lib == internal_module->handle)
+		return;
+
+	native_library_lock ();
+
+	module = netcore_handle_lookup (lib);
+	if (!module)
+		goto leave;
+
+	ref_count = mono_refcount_dec (module);
+	if (ref_count > 0)
+		goto leave;
+
+	g_hash_table_remove (native_library_module_map, module);
+	g_hash_table_add (native_library_module_blacklist, module);
+	mono_dl_close (module);
+
+leave:
+	native_library_unlock ();
+}
+
+gpointer
+ves_icall_System_Runtime_InteropServices_NativeLibrary_GetSymbol (gpointer lib, MonoStringHandle symbol_name_handle, MonoBoolean throw_on_error, MonoError *error)
+{
+	MonoDl *module;
+	gpointer symbol = NULL;
+	char *symbol_name;
+
+	ERROR_LOCAL_BEGIN (local_error, error, throw_on_error)
+
+	symbol_name = mono_string_handle_to_utf8 (symbol_name_handle, error);
+	goto_if_nok (error, leave);
```

Comment:
Unlocks without lock in error path.  C++ destructors help avoid this common mistake.

Reply:
Oops, good catch.